File: .agents/skills/bmad/scripts/jev_recommend.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# ///
"""Advisory Jev workflow recommendation (opt-in, disabled by default).

Narrow judgment over supplied evidence: the caller (a BMad skill) derives
candidate skill ids from the installed registry, passes them here, and this
script either honors an explicit user choice, handles single-candidate cases
deterministically, or asks Jev through OpenRouter when semantic ambiguity
remains. Output is a JSON advisory signal on stdout; the calling skill keeps
full control and must fall back to its ordinary reasoning whenever the status
is not "ok".

Disabled by default. Set the decision-assist mode to ``suggest`` (advisory
recommendations) or ``shadow`` (evaluation only) via
``BMAD_DECISION_ASSIST_MODE`` or the ``[jev] mode`` central-config key. When
off, this makes zero network calls.
"""

from __future__ import annotations

import argparse
import json
import sys
from pathlib import Path

sys.dont_write_bytecode = True

try:
    from jev_adapter import JevClient, load_settings
    from jev_policy import (
        PolicyError,
        apply_request_integrity,
        build_integrity_questions,
        build_recommend_questions,
        build_state,
        interpret_recommend,
        parse_evidence,
        sanitize_candidate_id,
    )
except ModuleNotFoundError as error:
    if error.name not in ("tomllib", "jev_adapter", "jev_policy"):
        raise
    sys.stderr.write("error: run this script from its own directory (sibling modules missing).\n")
    raise SystemExit(3) from None


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Advisory Jev workflow recommendation")
    parser.add_argument("--request", required=True, help="the user's request text")
    parser.add_argument("--project-root", default=".", help="BMad project root for central config")
    parser.add_argument(
        "--candidates",
        required=True,
        help="comma-separated candidate skill ids derived from the installed registry",
    )
    parser.add_argument(
        "--chosen",
        default=None,
        help="explicit user choice: skip the provider entirely and echo the choice",
    )
    parser.add_argument(
        "--evidence",
        action="append",
        default=[],
        metavar="KEY=VALUE",
        help="bounded evidence item (repeatable)",
    )
    return parser


def main() -> int:
    args = build_parser().parse_args()
    project_root = Path(args.project_root).resolve()
    settings = load_settings(project_root)

    if not settings.callable:
        status = "disabled" if settings.mode == "off" else "unavailable"
        reason = "disabled_by_config" if status == "disabled" else "missing_openrouter_api_key"
        print(json.dumps({"status": status, "reason": reason, "calls_made": 0}))
        return 0

    chosen = None
    if args.chosen:
        try:
            chosen = sanitize_candidate_id(args.chosen)
        except PolicyError as error:
            sys.stderr.write(f"error: {error}\n")
            return 2
        # An explicit user choice outranks any model result: echo it and
        # skip the provider entirely (zero network calls).
        print(
            json.dumps(
                {"status": "ok", "source": "explicit_user_choice", "recommendation": {"id": chosen}, "calls_made": 0}
            )
        )
        return 0

    try:
        candidates = [sanitize_candidate_id(raw) for raw in args.candidates.split(",") if raw.strip()]
    except PolicyError as error:
        sys.stderr.write(f"error: {error}\n")
        return 2
    # Dedupe, preserving order, and cap the option count via policy checks.
    candidates = list(dict.fromkeys(candidates))

    if len(candidates) == 1:
        print(
            json.dumps(
                {"status": "ok", "source": "single_candidate", "recommendation": {"id": candidates[0]}, "calls_made": 0}
            )
        )
        return 0
    if len(candidates) == 0:
        print(json.dumps({"status": "unavailable", "reason": "no_candidates", "calls_made": 0}))
        return 0

    try:
        evidence = parse_evidence(args.evidence)
    except PolicyError as error:
        sys.stderr.write(f"error: {error}\n")
        return 2

    questions = build_recommend_questions(candidates)
    state = build_state(args.request, evidence, max_chars=settings.max_state_chars)
    client = JevClient(settings)
    result = client.post_decision(
        operation="workflow_recommendation",
        state=state,
        questions=questions,
    )
    outcome = interpret_recommend(result, candidates, None)
    if outcome["status"] == "ok":
        # Second-stage prompt-injection gate: a genuine serial dependency
        # (it only matters once every recommendation gate already passed).
        integrity_result = client.post_decision(
            operation="request_integrity",
            state=state,
            questions=build_integrity_questions(),
        )
        outcome = apply_request_integrity(outcome, integrity_result)
    outcome["calls_made"] = client.calls_used
    if result.status == "ok":
        outcome["usage"] = result.usage
        outcome["model"] = result.model
    outcome["mode"] = settings.mode
    outcome["advisory"] = True
    outcome["confidence_note"] = (
        "confidence is the provider's distribution-concentration statistic, not a probability of correctness"
    )
    print(json.dumps(outcome))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: _bmad/custom/bmad-jev/jev_recommend.ts
#!/usr/bin/env -S npx tsx
/**
 * Advisory Jev workflow recommendation (opt-in, disabled by default).
 *
 * TypeScript port of jev_recommend.py. Narrow judgment over supplied
 * evidence: the caller (a BMad skill) derives candidate skill ids from the
 * installed registry, passes them here, and this script either honors an
 * explicit user choice, handles single-candidate cases deterministically,
 * or asks Jev when semantic ambiguity remains. Output is a JSON advisory
 * signal on stdout; the calling skill keeps full control and must fall back
 * to its ordinary reasoning whenever the status is not "ok".
 *
 * Disabled by default. Set the decision-assist mode to `suggest` or
 * `shadow` via BMAD_DECISION_ASSIST_MODE or the `[jev] mode` central-config
 * key. When off, this makes zero network calls.
 */

import { parseArgs } from "node:util";
import { resolve } from "node:path";
import { JevClient, loadSettings } from "./jev_adapter.ts";
import {
  PolicyError,
  pyDumps,
  tagPythonFloats,
  applyRequestIntegrity,
  buildIntegrityQuestions,
  buildRecommendQuestions,
  buildState,
  interpretRecommend,
  parseEvidence,
  sanitizeCandidateId,
  type JsonTable,
} from "./jev_policy.ts";

function emitJson(obj: JsonTable): void {
  // JSON on stdout, no tracebacks (contract). Serialized with Python
  // json.dumps-compatible spacing so Python- and TS-era outputs stay
  // byte-identical for golden tests.
  process.stdout.write(pyDumps(tagPythonFloats(obj) as import("./jev_policy.ts").Json) + "\n");
}

/** Caller-input errors are bad_request, not outages: JSON on stdout, exit 2. */
function failUsage(reason: string): number {
  emitJson({ status: "bad_request", reason_kind: "usage", reason, calls_made: 0 });
  return 2;
}

const USAGE = `usage: npx tsx jev_recommend.ts --request REQUEST [--project-root ROOT]
                  --candidates IDS [--chosen ID] [--evidence KEY=VALUE]...`;

interface ParsedArgs {
  request: string;
  projectRoot: string;
  candidates: string;
  chosen: string | undefined;
  evidence: string[];
}

function parseCli(argv: string[]): ParsedArgs {
  let parsed;
  try {
    parsed = parseArgs({
      args: argv,
      options: {
        request: { type: "string", short: "r" },
        "project-root": { type: "string", default: "." },
        candidates: { type: "string", short: "c" },
        chosen: { type: "string" },
        evidence: { type: "string", multiple: true },
      },
      strict: true,
    });
  } catch (error) {
    process.stderr.write(`${(error as Error).message}\n${USAGE}\n`);
    process.exit(2);
  }
  const values = parsed.values as {
    request?: string;
    "project-root"?: string;
    candidates?: string;
    chosen?: string;
    evidence?: string[];
  };
  if (!values.request || !values.candidates) {
    process.stderr.write(`${USAGE}\n`);
    process.exit(2);
  }
  return {
    request: values.request,
    projectRoot: values["project-root"] ?? ".",
    candidates: values.candidates,
    chosen: values.chosen,
    evidence: values.evidence ?? [],
  };
}

async function main(): Promise<number> {
  const args = parseCli(process.argv.slice(2));
  const projectRoot = resolve(args.projectRoot);
  const settings = loadSettings(projectRoot);

  if (!(settings.mode !== "off" && Boolean(settings.apiKey))) {
    const status = settings.mode === "off" ? "disabled" : "unavailable";
    const reason = status === "disabled" ? "disabled_by_config" : "missing_api_key";
    const kind = status === "disabled" ? "disabled" : "missing_api_key";
    emitJson({ status, reason_kind: kind, reason, calls_made: 0 });
    return 0;
  }

  if (args.chosen !== undefined && args.chosen !== "") {
    let chosen: string;
    try {
      chosen = sanitizeCandidateId(args.chosen);
    } catch (error) {
      if (error instanceof PolicyError) return failUsage(error.message);
      throw error;
    }
    // An explicit user choice outranks any model result: echo it and
    // skip the provider entirely (zero network calls).
    emitJson({ status: "ok", source: "explicit_user_choice", recommendation: { id: chosen }, calls_made: 0 });
    return 0;
  }

  let candidates: string[];
  try {
    candidates = args.candidates
      .split(",")
      .filter((raw) => raw.trim())
      .map((raw) => sanitizeCandidateId(raw));
  } catch (error) {
    if (error instanceof PolicyError) return failUsage(error.message);
    throw error;
  }
  // Dedupe, preserving order, and cap the option count via policy checks.
  candidates = [...new Set(candidates)];

  if (candidates.length === 1) {
    emitJson({ status: "ok", source: "single_candidate", recommendation: { id: candidates[0] }, calls_made: 0 });
    return 0;
  }
  if (candidates.length === 0) return failUsage("no_candidates");

  let evidence: Record<string, string>;
  try {
    evidence = parseEvidence(args.evidence);
  } catch (error) {
    if (error instanceof PolicyError) return failUsage(error.message);
    throw error;
  }

  let questions;
  let state;
  try {
    questions = buildRecommendQuestions(candidates);
    state = buildState(args.request, evidence, settings.maxStateChars);
  } catch (error) {
    // e.g. >MAX_CANDIDATES: a caller error, not an outage — never a traceback
    if (error instanceof PolicyError) return failUsage(error.message);
    throw error;
  }
  const client = new JevClient(settings);
  const result = await client.postDecision({
    operation: "workflow_recommendation",
    state,
    questions,
  });
  let outcome = interpretRecommend(result, candidates, null);
  if (outcome["status"] === "ok") {
    // Second-stage prompt-injection gate: a genuine serial dependency
    // (it only matters once every recommendation gate already passed).
    const integrityResult = await client.postDecision({
      operation: "request_integrity",
      state,
      questions: buildIntegrityQuestions(),
    });
    outcome = applyRequestIntegrity(outcome, integrityResult);
  }
  outcome["calls_made"] = client.calls_used;
  if (result.status === "ok") {
    outcome["usage"] = result.usage;
    outcome["model"] = result.model ?? null;
  }
  outcome["mode"] = settings.mode;
  outcome["advisory"] = true;
  outcome["confidence_note"] =
    "confidence is the provider's distribution-concentration statistic, not a probability of correctness";
  emitJson(outcome);
  return 0;
}

main().then((code) => {
  process.exitCode = code;
});
